service-wfs3/src/main/java/org/oskari/service/wfs3/FilterToOAPIFCoreQuery.java

```diff
@@ -0,0 +1,443 @@
+import org.opengis.filter.PropertyIsLessThanOrEqualTo;
+import org.opengis.filter.PropertyIsLike;
+import org.opengis.filter.PropertyIsNil;
+import org.opengis.filter.PropertyIsNotEqualTo;
+import org.opengis.filter.PropertyIsNull;
+import org.opengis.filter.expression.Add;
+import org.opengis.filter.expression.Divide;
+import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.ExpressionVisitor;
+import org.opengis.filter.expression.Function;
+import org.opengis.filter.expression.Literal;
+import org.opengis.filter.expression.Multiply;
+import org.opengis.filter.expression.NilExpression;
+import org.opengis.filter.expression.PropertyName;
+import org.opengis.filter.expression.Subtract;
+import org.opengis.filter.spatial.BBOX;
+import org.opengis.filter.spatial.Beyond;
+import org.opengis.filter.spatial.Contains;
+import org.opengis.filter.spatial.Crosses;
+import org.opengis.filter.spatial.DWithin;
+import org.opengis.filter.spatial.Disjoint;
+import org.opengis.filter.spatial.Equals;
+import org.opengis.filter.spatial.Intersects;
+import org.opengis.filter.spatial.Overlaps;
+import org.opengis.filter.spatial.Touches;
+import org.opengis.filter.spatial.Within;
+import org.opengis.filter.temporal.After;
+import org.opengis.filter.temporal.AnyInteracts;
+import org.opengis.filter.temporal.Before;
+import org.opengis.filter.temporal.Begins;
+import org.opengis.filter.temporal.BegunBy;
+import org.opengis.filter.temporal.During;
+import org.opengis.filter.temporal.EndedBy;
+import org.opengis.filter.temporal.Ends;
+import org.opengis.filter.temporal.Meets;
+import org.opengis.filter.temporal.MetBy;
+import org.opengis.filter.temporal.OverlappedBy;
+import org.opengis.filter.temporal.TContains;
+import org.opengis.filter.temporal.TEquals;
+import org.opengis.filter.temporal.TOverlaps;
+
+import fi.nls.oskari.domain.map.OskariLayer;
+import fi.nls.oskari.util.JSONHelper;
+
+public class FilterToOAPIFCoreQuery implements FilterVisitor, ExpressionVisitor {
+
+    static final String ATTRIBUTE_QUERYABLES = "queryables";
+
+    private final OskariLayer layer;
+    private boolean insideAnd = false;
+
+    public FilterToOAPIFCoreQuery(OskariLayer layer) {
+        this.layer = layer;
+    }
+
+    public Filter toQueryParameters(Filter filter, Map<String, String> query) {
+        FilterCapabilities capabilities = createFilterCapabilities(layer);
+        Set<String> queryables = getQueryables(layer);
+
+        SimpleAndFlatteningFilterVisitor flattener = new SimpleAndFlatteningFilterVisitor(CommonFactoryFinder.getFilterFactory());
+        filter = (Filter) filter.accept(flattener, null);
+
+        OAPIFCoreFilterSplittingVisitor splitter = new OAPIFCoreFilterSplittingVisitor(capabilities, queryables);
+        filter.accept(splitter, null);
+
+        Filter preFilter = splitter.getFilterPre();
+        preFilter = (Filter) preFilter.accept(flattener, null);
+        preFilter.accept(this, query);
```

Comment:
What's the logic behind sometimes assigning variable with object returned by accept() and sometimes not? Does `accept()` mutate it's input or return a new instance or mutates and returns the mutated instance? I guess it both mutates and returns the original instance? It would be easier to read if the return value is either always or never assigned. If there's a reason maybe add a comment about it?

Reply:
The behaviour depends on the implementation of FilterVisitor. I'll just get rid of `SimpleAndFlatteningFilterVisitor` as it's not readable and is too complex for the use-case which boils down to one recursive functions with couple of instanceof checks.

Reply:
See https://github.com/oskariorg/oskari-server/pull/816/commits/b2951d8788daf364fc60fed847ed80f434f84829